axiom/client.go

```diff
@@ -335,12 +379,67 @@ func (c *Client) do(req *http.Request, v interface{}) (*response, error) {
 			return resp, err
 		}
 
-		dec := json.NewDecoder(resp.Body)
-		if c.strictDecoding {
-			dec.DisallowUnknownFields()
+		if val := resp.Header.Get(headerContentType); strings.HasPrefix(val, mediaTypeJSON) {
+			dec := json.NewDecoder(resp.Body)
+			if c.strictDecoding {
+				dec.DisallowUnknownFields()
+			}
+			return resp, dec.Decode(v)
 		}
-		return resp, dec.Decode(v)
+
+		return resp, errors.New("cannot decode response with unknown content type")
 	}
 
 	return resp, nil
 }
+
+// checkLimit checks if *LimitError can be immediately returned from
+// `Client.do`, and if so, returns it so that `Client.do` can skip making an API
+// call unnecessarily.
+func (c *Client) checkLimit(req *http.Request) *LimitError {
+	var (
+		lt            limitType
+		messagePrefix string
+	)
+	if strings.HasSuffix(req.URL.Path, "/ingest") {
+		lt = limitIngest
+		messagePrefix = "ingest"
+	} else if strings.HasSuffix(req.URL.Path, "/query") || strings.HasSuffix(req.URL.Path, "/_apl") {
+		lt = limitQuery
+		messagePrefix = "query"
+	} else {
+		lt = limitRate
+		messagePrefix = "rate"
+	}
+
+	var limit Limit
+	c.limitsMu.Lock()
+	for ls := LimitScopeUnknown; ls <= LimitScopeAnonymous; ls++ {
+		key := limitKey(lt, ls)
+		var ok bool
+		if limit, ok = c.limits[key]; ok {
+			break
+		}
+	}
+	c.limitsMu.Unlock()
+
+	if !c.noLimiting && !limit.Reset.IsZero() && limit.Remaining == 0 && time.Now().Before(limit.Reset) {
+		// Create a fake response.
+		resp := &http.Response{
+			Status:     http.StatusText(http.StatusTooManyRequests),
+			StatusCode: http.StatusTooManyRequests,
+			Request:    req,
+			Header:     make(http.Header),
+			Body:       ioutil.NopCloser(strings.NewReader("")),
```

Comment:
Can we reuse this fake body?

Reply:
Like `var fakeBody = ioutil.NopCloser(strings.NewReader(""))` and pass `faceBody` around?

Probably.

Reply:
Yeah probably fine without